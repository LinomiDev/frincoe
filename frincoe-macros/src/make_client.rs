use proc_macro2::TokenStream;
use quote::quote;
use syn::parse::{Parse, ParseStream};
use syn::{Ident, LitStr, Path, Token, TraitItem, Type};

use crate::helpers::read_trait;

struct ClientArgs {
    pub adapter: Ident,
    pub filename: LitStr,
    pub srcpath: Path,
    pub modpath: Option<Path>,
    pub target: Type,
}

impl Parse for ClientArgs {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        // impl "path"::mod::trait ...
        input.parse::<Token![impl]>()?;
        let filename: LitStr = input.parse()?;
        input.parse::<Token![::]>()?;
        let srcpath: Path = input.parse()?;
        // [as mod::trait] ...
        let lookahead = input.lookahead1();
        let modpath: Option<Path> = if lookahead.peek(Token![as]) {
            input.parse::<Token![as]>()?;
            Some(input.parse()?)
        } else {
            None
        };
        // for Type ...
        input.parse::<Token![for]>()?;
        let target: Type = input.parse()?;
        // in adapter
        input.parse::<Token![in]>()?;
        let adapter: Ident = input.parse()?;
        // )
        if !input.is_empty() {
            return Err(input.error("Wrong syntax for make_client: nothing should be following the struct"));
        }
        Ok(Self {
            adapter,
            filename,
            srcpath,
            modpath,
            target,
        })
    }
}

pub fn make_client_impl(args: TokenStream) -> TokenStream {
    let ClientArgs {
        adapter,
        filename,
        srcpath,
        modpath,
        target,
    } = match syn::parse2(args) {
        Ok(v) => v,
        Err(err) => return err.to_compile_error(),
    };

    // Generate the injecting content
    let content = match read_trait(filename, &srcpath) {
        Some(v) => v,
        None => return TokenStream::new(),
    }
    .iter()
    .map(|item| match item {
        TraitItem::Macro(_) => panic!("macros in trait declaration are not currently supported"),
        value => quote! { #adapter !(#value); },
    })
    .collect::<Vec<_>>();
    let modpath = modpath.unwrap_or(srcpath);

    quote! {
        impl #modpath for #target {
            #(#content)*
        }
    }
}
