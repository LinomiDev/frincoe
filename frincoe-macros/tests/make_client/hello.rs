use frincoe_macros::make_client;

include!("interfaces.rs");

macro_rules! Error {
    ( fn $func:ident $args:tt -> $ret:ty; ) => {
        compile_error!(concat!(
            stringify!($func),
            stringify!($args),
            " -> ",
            stringify!($ret)
        ));
    };
}

struct Target {}

make_client!(impl "interfaces.rs"::SayHello for Target in Error);

fn main() {}
