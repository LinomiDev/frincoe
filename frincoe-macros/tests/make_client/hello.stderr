error: hello(& self, name : & str) -> std::io::Result<()>
  --> tests/make_client/hello.rs:7:9
   |
7  | /         compile_error!(concat!(
8  | |             stringify!($func),
9  | |             stringify!($args),
10 | |             " -> ",
11 | |             stringify!($ret)
12 | |         ));
   | |__________^
...
18 |   make_client!(impl "interfaces.rs"::SayHello for Target in Error);
   |   ---------------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `Error` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0046]: not all trait items implemented, missing: `hello`
  --> tests/make_client/hello.rs:18:1
   |
18 | make_client!(impl "interfaces.rs"::SayHello for Target in Error);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `hello` in implementation
   |
  ::: tests/make_client/interfaces.rs
   |
   |     fn hello(&self, name: &str) -> std::io::Result<()>;
   |     --------------------------------------------------- `hello` from trait
   |
   = note: this error originates in the macro `make_client` (in Nightly builds, run with -Z macro-backtrace for more info)
